### PR TITLE
Allows negative index to tensor.narrow (Fixes: #9546)

### PR DESCRIPTION
**Pull request by bhushan23**
Fixes #9546 
Test cases added
reviewed by:

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -126,7 +126,7 @@ Tensor narrow(const Tensor& self, int64_t dim, int64_t start, int64_t length) {
   AT_CHECK(self.dim() > 0, "narrow() cannot be applied to a 0-dim tensor.");
   auto cur_size = self.size(dim);
   if (start < 0) {
-    AT_ERROR("start out of range");
+    start = cur_size + start;
```
Comment (zou3519) [inline aten/src/ATen/native/TensorShape.cpp]: let's say start is a very, very small number (-1000000). Then start would be out of bounds after this line and we wouldn't be able to report the error here to the user.

You should try to call `maybe_wrap_dim` [(usage here)](https://github.com/pytorch/pytorch/blob/64834f6fb8055cf2d633353992288a8823b915ed/aten/src/ATen/native/TensorShape.cpp#L52) instead

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -126,7 +126,7 @@ Tensor narrow(const Tensor& self, int64_t dim, int64_t start, int64_t length) {
   AT_CHECK(self.dim() > 0, "narrow() cannot be applied to a 0-dim tensor.");
   auto cur_size = self.size(dim);
   if (start < 0) {
```
Comment (vishwakftw) [inline aten/src/ATen/native/TensorShape.cpp]: I think this condition must be changed to `AT_CHECK(-cur_size <= start < cur_size, "start out of range")`, and then an `if` case for `start < 0 `.

File: test/test_torch.py [hunk 3]
```diff
@@ -3424,6 +3424,16 @@ def _test_cat_empty(self, use_cuda=False):
 
     def test_cat_empty(self):
         self._test_cat_empty(self)
+    
+    def test_narrow(self):
+        x = torch.Tensor([ [0, 1, 2], [3, 4, 5], [6, 7, 8] ])
```
Comment (vishwakftw) [inline test/test_torch.py]: Please remove the white-space around outer `[` and `]`.

File: test/test_torch.py [hunk 4]
```diff
@@ -3424,6 +3424,16 @@ def _test_cat_empty(self, use_cuda=False):
 
     def test_cat_empty(self):
         self._test_cat_empty(self)
+
```
Comment (vishwakftw) [inline test/test_torch.py]: White-space needs to be removed.

File: aten/src/ATen/native/TensorShape.cpp (hunk 2 above)
Comment (vishwakftw) [reply aten/src/ATen/native/TensorShape.cpp]: Alternatively, you can use `maybe_wrap_dim` from `WrapDimUtils.h`.

**Review by ssnl — DISMISSED** [review]
test_autograd needs new test cases with negative start/end

**Comment by bhushan23** [comment]
@SsnL why do we need test cases for test_autograd? 
autograd behavior is independent of what narrow will output, right? how are these two related? maybe I am missing something!

Are you suggesting to replicate **test_inplace_view_backprop_base** and add narrow with negative start?

**Comment by SsnL** [comment]
nvm, it's fine because it calls select and the backward is from select. but why is only start wrap_dim'ed?

**Comment by bhushan23** [comment]
I though tensor size should take care of it and hence dim don't need it. 
Anyways, added for dim as well

**Comment by SsnL** [comment]
Nah you are right, `dim` doesn't need it because it is handled in `select`. My bad. Could you revert that change since it's inducing an extra unnecessary check? Sorry about it.

**Comment by bhushan23** [comment]
done

**Comment by bhushan23** [comment]
@SsnL @zou3519

**Review by zou3519 — APPROVED** [review]
lgtm, thanks @bhushan23

**Comment by bhushan23** [comment]
@ezyang please review and approve

**Comment by vishwakftw** [comment]
This can be closed now, it's merged in.

**Comment by gchanan** [comment]
FYI this isn't quite correct; it's (sometimes) valid to specify that the start is equal to the size of a dimension for narrow, index, etc., but that is not a valid dimension specification.  See https://github.com/pytorch/pytorch/pull/9838.

**Pull request closed without merging**